test/performance/benchmarks/real-traffic-test/main.go

```diff
@@ -0,0 +1,301 @@
+
+	// Test configuration
+	// Defines the latency of target
+	minLatency = 0 * time.Second
+	maxLatency = 5 * time.Second
+
+	// Defines the delay that a Knative Service has for startup (init-container causes the delay)
+	minStartupLatency = 0 * time.Second
+	maxStartupLatency = 10 * time.Second
+
+	// Defines the payloads that are sent on the vegeta requests
+	minPayloadSizeBytes = 10
+	maxPayloadSizeBytes = 50_000
+)
+
+var (
+	numberOfServices = flag.Int("number-of-services", 10, "The number of Knative Services to create")
+	rps              = flag.Int("requests-per-second", 300, "The of requests per second to send")
+)
+
+type serviceConfig struct {
+	resourceObjects *v1test.ResourceObjects
+
+	activatorAlwaysInPath bool
+	latency               int64
+	startupLatency        int64
+	payload               []byte
+}
+
+func main() {
+	if *rps >= 2500 {
+		log.Fatal("One test container cannot create more than 2500 RPS without errors. Consider starting this test in parallel.")
+	}
+
+	ctx := signals.NewContext()
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Minute)
+	defer cancel()
+
+	// To make testing.T work properly
+	testing.Init()
+
+	env := environment.ClientConfig{}
+
+	// The local domain is directly resolvable by the test
+	flag.Set("resolvabledomain", "true")
+
+	// manually parse flags to avoid conflicting flags
+	flag.Parse()
+
+	cfg, err := env.GetRESTConfig()
+	if err != nil {
+		log.Fatalf("failed to get kubeconfig %s", err)
+	}
+
+	ctx, _ = injection.EnableInjectionOrDie(ctx, cfg)
+
+	clients, err := test.NewClients(cfg, namespace)
+	if err != nil {
+		log.Fatal("Failed to setup clients: ", err)
+	}
+
+	influxReporter, err := performance.NewInfluxReporter(map[string]string{"number-of-services": strconv.Itoa(*numberOfServices)})
+	if err != nil {
+		log.Fatalf("failed to create influx reporter: %v", err.Error())
+	}
+	defer influxReporter.FlushAndShutdown()
+
+	log.Printf("Creating %d Knative Services", *numberOfServices)
+	services, cleanup, err := createServices(clients, *numberOfServices)
+	if err != nil {
+		log.Fatalf("Failed to create services: %v", err)
+	}
+	defer cleanup()
+
+	log.Print("Creating vegeta targets")
+
+	targets := []vegeta.Target{}
+	for _, svc := range services {
+		t := vegeta.Target{
+			Method: http.MethodPost,
+			URL:    fmt.Sprintf("http://%s.default.svc.cluster.local?sleep=%d", svc.resourceObjects.Service.Name, svc.latency),
+			Body:   svc.payload,
+		}
+		targets = append(targets, t)
+
+		log.Printf("Name: %s, startup-latency: %ds, latency: %dms, payload: %dKb, activator-always-in-path: %v",
+			svc.resourceObjects.Service.Name, svc.startupLatency, svc.latency, len(svc.payload)/1000, svc.activatorAlwaysInPath)
+	}
+
+	// Send configured RPS round-robin over all services,
+	// while the request timeout is based on the max delays + 20 seconds
+	log.Printf("Starting vegeta attack for with %v RPS for duration: %v", *rps, duration)
+	rate := vegeta.Rate{Freq: *rps, Per: time.Second}
```

Comment:
question: is vegeta able to report/error out if it is unable to maintain the desired RPS?

Reply:
I don't think so: https://github.com/tsenart/vegeta. But I think we could check as the actual rate is reported. But not sure if we have a target rate we want to be able to achieve?

Reply:
As a follow up it would be good to fail the test if we can't achieve the desired request rate. 

It would be good to catch if our jobs are underprovisioned.

Reply:
https://github.com/knative/serving/issues/14403